pkg/agent/packetcapture/packetcapture_controller_linux.go

```diff
@@ -0,0 +1,884 @@
+	c.queue.Add(pc.Name)
+}
+
+// Run will create defaultWorkers workers (go routines) which will process the PacketCapture events from the
+// workqueue.
+func (c *Controller) Run(stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+
+	klog.InfoS("Starting packetcapture controller", "name", controllerName)
+	defer klog.InfoS("Shutting down packetcapture controller", "name", controllerName)
+
+	cacheSynced := []cache.InformerSynced{c.packetCaptureSynced}
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, cacheSynced...) {
+		return
+	}
+
+	err := defaultFS.MkdirAll(packetDirectory, 0755)
+	if err != nil {
+		klog.ErrorS(err, "Couldn't create the directory for storing captured packets", "directory", packetDirectory)
+		return
+	}
+
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (c *Controller) addPacketCapture(obj interface{}) {
+	pc := obj.(*crdv1alpha1.PacketCapture)
+	klog.InfoS("Processing PacketCapture ADD event", "name", pc.Name)
+	c.enqueuePacketCapture(pc)
+}
+
+func (c *Controller) updatePacketCapture(_, obj interface{}) {
+	pc := obj.(*crdv1alpha1.PacketCapture)
+	klog.InfoS("Processing PacketCapture UPDATE event", "name", pc.Name)
+	c.enqueuePacketCapture(pc)
+}
+
+func (c *Controller) deletePacketCapture(obj interface{}) {
+	pc := obj.(*crdv1alpha1.PacketCapture)
+	klog.InfoS("Processing PacketCapture DELETE event", "name", pc.Name)
+	c.enqueuePacketCapture(pc)
+}
+
+func nameToPath(name string) string {
+	return filepath.Join(packetDirectory, name+".pcapng")
+}
+
+func (c *Controller) worker() {
+	go c.processWaitingCaptures()
+	for c.processPacketCaptureItem() {
+	}
+}
+
+func (c *Controller) processWaitingCaptures() {
+	for {
+		capture := func() string {
+			c.cond.L.Lock()
+			defer c.cond.L.Unlock()
+			for c.numCaptures >= maxConcurrentCaptures || c.waitingCaptures == 0 {
+				c.cond.Wait()
+			}
+			c.numCaptures += 1
+			c.waitingCaptures -= 1
+			capture := <-c.waitingCapturesCh
+			c.startPacketCapture(capture)
+			c.captures[capture] = packetCapturePhaseRunning
+			return capture
+		}()
+		c.queue.Add(capture)
+	}
+}
+
+func (c *Controller) processPacketCaptureItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+	if err := c.syncPacketCapture(key); err == nil {
+		c.queue.Forget(key)
+	} else {
+		c.queue.AddRateLimited(key)
+		klog.ErrorS(err, "Error syncing PacketCapture, requeueing", "key", key)
+	}
+	return true
+}
+
+func (c *Controller) cleanupPacketCapture(pcName string) {
+	path := nameToPath(pcName)
+	exist, err := afero.Exists(defaultFS, path)
+	if err != nil {
+		klog.ErrorS(err, "Failed to check if path exists", "path", path)
+	}
+	if !exist {
+		return
+	}
+	if err := defaultFS.Remove(path); err == nil {
+		klog.V(2).InfoS("Deleted pcap file", "name", pcName, "path", path)
+	} else {
+		klog.ErrorS(err, "Failed to delete pcap file", "name", pcName, "path", path)
```

Comment:
It seems you will return error directly in `getPacketFileAndWriter` if the file already exists, then how would you handle the case that a PacketCapture is deleted but the captured file is not deleted successfully, and later a new PacketCapture with the same name is created?

I suppose you should remove the existing one and recreate it in `getPacketFileAndWriter` instead of return an error? @antoninbas  any thoughts?

Reply:
i don't have any strong opinions on this.  will update this for now.